File: cloud/acp-spring-cloud-starter/src/main/java/pers/acp/spring/cloud/log/LogInstance.java
package pers.acp.spring.cloud.log;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.messaging.support.MessageBuilder;
import org.springframework.stereotype.Component;
import pers.acp.core.CommonTools;
import pers.acp.core.log.LogFactory;
import pers.acp.spring.boot.tools.SpringBeanFactory;
import pers.acp.spring.cloud.enums.LogLevel;
import pers.acp.spring.cloud.log.producer.LogProducer;
import pers.acp.spring.cloud.conf.LogServerClientConfiguration;

import java.util.Arrays;

/**
 * 日志实例
 *
 * @author zhangbin by 11/07/2018 13:36
 * @since JDK 11
 */
@Component
public class LogInstance {

    private final LogFactory log = LogFactory.getInstance(this.getClass());

    private final ObjectMapper objectMapper;

    private final LogServerClientConfiguration logServerClientConfiguration;

    @Autowired
    public LogInstance(ObjectMapper objectMapper, LogServerClientConfiguration logServerClientConfiguration) {
        this.objectMapper = objectMapper;
        this.logServerClientConfiguration = logServerClientConfiguration;
    }

    private LogInfo generateLogInfo() {
        LogInfo logInfo = SpringBeanFactory.getBean(LogInfo.class);
        if (logInfo != null) {
            String logType = logServerClientConfiguration.getLogType();
            if (CommonTools.isNullStr(logType)) {
                logType = LogConstant.DEFAULT_TYPE;
            }
            logInfo.setLogType(logType);
        }
        return logInfo;
    }

    private void sendToLogServer(LogInfo logInfo) {
        logInfo.setServerTime(CommonTools.getNowDateTime().toDate().getTime());
        StackTraceElement[] stacks = Thread.currentThread().getStackTrace();
        int lineno = 0;
        String className = "";
        if (stacks.length >= 4) {
            lineno = stacks[3].getLineNumber();
            className = stacks[3].getClassName();
        }
        logInfo.setLineno(lineno);
        logInfo.setClassName(className);
        try {
            if (logServerClientConfiguration.isEnabled()) {
                LogProducer logProducer = SpringBeanFactory.getBean(LogProducer.class);
                if (logProducer != null) {
                    logProducer.getLogOutput().sendMessage().send(MessageBuilder.withPayload(objectMapper.writeValueAsString(logInfo)).build());
                }
            }
        } catch (JsonProcessingException e) {
            log.error(e.getMessage(), e);
        }
    }

    public void info(String message) {
        log.info(message);
        LogInfo logInfo = generateLogInfo();
        if (logInfo != null) {
            logInfo.setLogLevel(LogLevel.INFO.getValue());
            logInfo.setMessage(message);
            sendToLogServer(logInfo);
        }
    }

    public void info(String message, Object... var) {
        log.info(message, var);
        LogInfo logInfo = generateLogInfo();
        if (logInfo != null) {
            logInfo.setLogLevel(LogLevel.INFO.getValue());
            logInfo.setMessage(message);
            logInfo.setParams(Arrays.asList(var));
            sendToLogServer(logInfo);
        }
    }

    public void info(String message, Throwable t) {
        log.info(message, t);
        LogInfo logInfo = generateLogInfo();
        if (logInfo != null) {
            logInfo.setLogLevel(LogLevel.INFO.getValue());
            logInfo.setMessage(message);
            logInfo.setThrowable(t);
            sendToLogServer(logInfo);
        }
    }

    public void debug(String message) {
        log.debug(message);
        LogInfo logInfo = generateLogInfo();
        if (logInfo != null) {
            logInfo.setLogLevel(LogLevel.DEBUG.getValue());
            logInfo.setMessage(message);
            sendToLogServer(logInfo);
        }
    }

    public void debug(String message, Object... var) {
        log.debug(message, var);
        LogInfo logInfo = generateLogInfo();
        if (logInfo != null) {
            logInfo.setLogLevel(LogLevel.DEBUG.getValue());
            logInfo.setMessage(message);
            logInfo.setParams(Arrays.asList(var));
            sendToLogServer(logInfo);
        }
    }

    public void debug(String message, Throwable t) {
        log.debug(message, t);
        LogInfo logInfo = generateLogInfo();
        if (logInfo != null) {
            logInfo.setLogLevel(LogLevel.DEBUG.getValue());
            logInfo.setMessage(message);
            logInfo.setThrowable(t);
            sendToLogServer(logInfo);
        }
    }

    public void warn(String message) {
        log.warn(message);
        LogInfo logInfo = generateLogInfo();
        if (logInfo != null) {
            logInfo.setLogLevel(LogLevel.WARN.getValue());
            logInfo.setMessage(message);
            sendToLogServer(logInfo);
        }
    }

    public void warn(String message, Object... var) {
        log.warn(message, var);
        LogInfo logInfo = generateLogInfo();
        if (logInfo != null) {
            logInfo.setLogLevel(LogLevel.WARN.getValue());
            logInfo.setMessage(message);
            logInfo.setParams(Arrays.asList(var));
            sendToLogServer(logInfo);
        }
    }

    public void warn(String message, Throwable t) {
        log.warn(message, t);
        LogInfo logInfo = generateLogInfo();
        if (logInfo != null) {
            logInfo.setLogLevel(LogLevel.WARN.getValue());
            logInfo.setMessage(message);
            logInfo.setThrowable(t);
            sendToLogServer(logInfo);
        }
    }

    public void error(String message) {
        log.error(message);
        LogInfo logInfo = generateLogInfo();
        if (logInfo != null) {
            logInfo.setLogLevel(LogLevel.ERROR.getValue());
            logInfo.setMessage(message);
            sendToLogServer(logInfo);
        }
    }

    public void error(String message, Object... var) {
        log.error(message, var);
        LogInfo logInfo = generateLogInfo();
        if (logInfo != null) {
            logInfo.setLogLevel(LogLevel.ERROR.getValue());
            logInfo.setMessage(message);
            logInfo.setParams(Arrays.asList(var));
            sendToLogServer(logInfo);
        }
    }

    public void error(String message, Throwable t) {
        log.error(message, t);
        LogInfo logInfo = generateLogInfo();
        if (logInfo != null) {
            logInfo.setLogLevel(LogLevel.ERROR.getValue());
            logInfo.setMessage(message);
            logInfo.setThrowable(t);
            sendToLogServer(logInfo);
        }
    }

    public void trace(String message) {
        log.trace(message);
        LogInfo logInfo = generateLogInfo();
        if (logInfo != null) {
            logInfo.setLogLevel(LogLevel.TRACE.getValue());
            logInfo.setMessage(message);
            sendToLogServer(logInfo);
        }
    }

    public void trace(String message, Object... var) {
        log.trace(message, var);
        LogInfo logInfo = generateLogInfo();
        if (logInfo != null) {
            logInfo.setLogLevel(LogLevel.TRACE.getValue());
            logInfo.setMessage(message);
            logInfo.setParams(Arrays.asList(var));
            sendToLogServer(logInfo);
        }
    }

    public void trace(String message, Throwable t) {
        log.trace(message, t);
        LogInfo logInfo = generateLogInfo();
        if (logInfo != null) {
            logInfo.setLogLevel(LogLevel.TRACE.getValue());
            logInfo.setMessage(message);
            logInfo.setThrowable(t);
            sendToLogServer(logInfo);
        }
    }

}

File: cloud/acp-spring-cloud-starter/src/main/kotlin/pers/acp/spring/cloud/log/LogInstance.kt
package pers.acp.spring.cloud.log

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.messaging.support.MessageBuilder
import org.springframework.stereotype.Component
import pers.acp.core.CommonTools
import pers.acp.core.log.LogFactory
import pers.acp.spring.boot.tools.SpringBeanFactory
import pers.acp.spring.cloud.enums.LogLevel
import pers.acp.spring.cloud.log.producer.LogProducer
import pers.acp.spring.cloud.conf.LogServerClientConfiguration

/**
 * 日志实例
 *
 * @author zhangbin by 11/07/2018 13:36
 * @since JDK 11
 */
@Component
class LogInstance @Autowired
constructor(private val objectMapper: ObjectMapper, private val logServerClientConfiguration: LogServerClientConfiguration) {

    private val log = LogFactory.getInstance(this.javaClass)

    private fun generateLogInfo(): LogInfo? {
        val logInfo = SpringBeanFactory.getBean(LogInfo::class.java)
        logInfo?.let {
            var logType = logServerClientConfiguration.logType
            if (CommonTools.isNullStr(logType)) {
                logType = LogConstant.DEFAULT_TYPE
            }
            it.logType = logType
        }
        return logInfo
    }

    private fun sendToLogServer(logInfo: LogInfo) {
        logInfo.serverTime = CommonTools.getNowDateTime().toDate().time
        val stacks = Thread.currentThread().stackTrace
        var lineno = 0
        var className = ""
        if (stacks.size >= 4) {
            lineno = stacks[3].lineNumber
            className = stacks[3].className
        }
        logInfo.lineno = lineno
        logInfo.className = className
        try {
            if (logServerClientConfiguration.isEnabled) {
                val logProducer = SpringBeanFactory.getBean(LogProducer::class.java)
                logProducer?.logOutput?.sendMessage()?.send(MessageBuilder.withPayload(objectMapper.writeValueAsString(logInfo)).build())
            }
        } catch (e: JsonProcessingException) {
            log.error(e.message, e)
        }

    }

    fun info(message: String) {
        log.info(message)
        val logInfo = generateLogInfo()
        logInfo?.let {
            it.logLevel = LogLevel.Info.name
            it.message = message
            sendToLogServer(it)
        }
    }

    fun info(message: String, vararg variable: Any?) {
        log.info(message, *variable)
        val logInfo = generateLogInfo()
        logInfo?.let {
            it.logLevel = LogLevel.Info.name
            it.message = message
            it.params = arrayListOf(*variable)
            sendToLogServer(it)
        }
    }

    fun info(message: String, t: Throwable) {
        log.info(message, t)
        val logInfo = generateLogInfo()
        logInfo?.let {
            it.logLevel = LogLevel.Info.name
            it.message = message
            it.throwable = t
            sendToLogServer(it)
        }
    }

    fun debug(message: String) {
        log.debug(message)
        val logInfo = generateLogInfo()
        logInfo?.let {
            it.logLevel = LogLevel.Debug.name
            it.message = message
            sendToLogServer(it)
        }
    }

    fun debug(message: String, vararg variable: Any?) {
        log.debug(message, *variable)
        val logInfo = generateLogInfo()
        logInfo?.let {
            it.logLevel = LogLevel.Debug.name
            it.message = message
            it.params = arrayListOf(*variable)
            sendToLogServer(it)
        }
    }

    fun debug(message: String, t: Throwable) {
        log.debug(message, t)
        val logInfo = generateLogInfo()
        logInfo?.let {
            it.logLevel = LogLevel.Debug.name
            it.message = message
            it.throwable = t
            sendToLogServer(it)
        }
    }

    fun warn(message: String) {
        log.warn(message)
        val logInfo = generateLogInfo()
        logInfo?.let {
            it.logLevel = LogLevel.Warn.name
            it.message = message
            sendToLogServer(it)
        }
    }

    fun warn(message: String, vararg variable: Any?) {
        log.warn(message, *variable)
        val logInfo = generateLogInfo()
        logInfo?.let {
            it.logLevel = LogLevel.Warn.name
            it.message = message
            it.params = arrayListOf(*variable)
            sendToLogServer(it)
        }
    }

    fun warn(message: String, t: Throwable) {
        log.warn(message, t)
        val logInfo = generateLogInfo()
        logInfo?.let {
            it.logLevel = LogLevel.Warn.name
            it.message = message
            it.throwable = t
            sendToLogServer(it)
        }
    }

    fun error(message: String) {
        log.error(message)
        val logInfo = generateLogInfo()
        logInfo?.let {
            it.logLevel = LogLevel.Error.name
            it.message = message
            sendToLogServer(it)
        }
    }

    fun error(message: String, vararg variable: Any?) {
        log.error(message, *variable)
        val logInfo = generateLogInfo()
        logInfo?.let {
            it.logLevel = LogLevel.Error.name
            it.message = message
            it.params = arrayListOf(*variable)
            sendToLogServer(it)
        }
    }

    fun error(message: String, t: Throwable) {
        log.error(message, t)
        val logInfo = generateLogInfo()
        logInfo?.let {
            it.logLevel = LogLevel.Error.name
            it.message = message
            it.throwable = t
            sendToLogServer(it)
        }
    }

    fun trace(message: String) {
        log.trace(message)
        val logInfo = generateLogInfo()
        logInfo?.let {
            it.logLevel = LogLevel.Trace.name
            it.message = message
            sendToLogServer(it)
        }
    }

    fun trace(message: String, vararg variable: Any?) {
        log.trace(message, *variable)
        val logInfo = generateLogInfo()
        logInfo?.let {
            it.logLevel = LogLevel.Trace.name
            it.message = message
            it.params = arrayListOf(*variable)
            sendToLogServer(it)
        }
    }

    fun trace(message: String, t: Throwable) {
        log.trace(message, t)
        val logInfo = generateLogInfo()
        logInfo?.let {
            it.logLevel = LogLevel.Trace.name
            it.message = message
            it.throwable = t
            sendToLogServer(it)
        }
    }

}